<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><link rel="canonical" href="https://clinical-genomics.github.io/chanjo2/usage/coverage-reports/" />
      <link rel="shortcut icon" href="../../img/favicon.ico" />
    <title>Coverage and genes overview reports - Chanjo2</title>
    <link rel="stylesheet" href="../../css/theme.css" />
    <link rel="stylesheet" href="../../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "Coverage and genes overview reports";
        var mkdocs_page_input_path = "usage/coverage-reports.md";
        var mkdocs_page_url = "/chanjo2/usage/coverage-reports/";
      </script>
    
    <!--[if lt IE 9]>
      <script src="../../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
      <script>hljs.highlightAll();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href="../.." class="icon icon-home"> Chanjo2
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../..">Index</a>
                </li>
              </ul>
              <p class="caption"><span class="caption-text">Deployment</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../../deployment/docker-deployment/">Deployment using Docker</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../deployment/installation/">Installation on a virtual environment</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../../deployment/env_file/">Custom settings and the .env file</a>
                  </li>
              </ul>
              <p class="caption"><span class="caption-text">Usage</span></p>
              <ul class="current">
                  <li class="toctree-l1"><a class="reference internal" href="../loading-intervals/">Loading and displaying genes, transcripts and exons</a>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../sample-coverage/">Coverage and coverage completeness queries</a>
                  </li>
                  <li class="toctree-l1 current"><a class="reference internal current" href="#">Coverage and genes overview reports</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#mandatory-parameters">Mandatory parameters</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#optional-parameters">Optional parameters</a>
    </li>
    </ul>
                  </li>
                  <li class="toctree-l1"><a class="reference internal" href="../authorised_requests/">Authorised requests</a>
                  </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../..">Chanjo2</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../.." class="icon icon-home" aria-label="Docs"></a></li>
          <li class="breadcrumb-item">Usage</li>
      <li class="breadcrumb-item active">Coverage and genes overview reports</li>
    <li class="wy-breadcrumbs-aside">
          <a href="https://github.com/Clinical-Genomics/chanjo2/edit/master/docs/usage/coverage-reports.md" class="icon icon-github"> Edit on GitHub</a>
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <h1 id="creating-a-coverage-report-for-one-or-more-samples-based-on-a-list-of-genes">Creating a coverage report for one or more samples based on a list of genes</h1>
<p>An example of this report is shown by the demo report endpoint: http://0.0.0.0:8000/report/demo (requires all genes and gene transcripts in build GRCh37 loaded into the database).</p>
<p><img width="924" alt="image" src="https://github.com/Clinical-Genomics/chanjo2/assets/28093618/02fcfcc8-94ce-4344-bc02-fae4609810cb"></p>
<p>The statistics from the report above are computed using the transcripts intervals present in the provided genes.</p>
<p>Given an application running in production settings, with <strong>genes, transcripts and exons</strong> loaded into the database, a coverage report like the one above can be created taking into account one of these types of intervals from any gene of choice.</p>
<p>The gene list might be provided as a list of <strong>Ensembl gene IDs, HGNC gene IDs or HGNC gene symbols</strong>.</p>
<p>Here is what the request data of a <strong>POST request to the <code>/report</code> endpoint</strong> looks like:</p>
<pre><code class="language-shell">{
  &quot;build&quot;: &quot;GRCh37&quot;,
  &quot;completeness_thresholds&quot;: [
    10,
    15,
    20,
    50,
    100
  ],
  &quot;ensembl_gene_ids&quot;: [],
  &quot;hgnc_gene_ids&quot;: [],
  &quot;hgnc_gene_symbols&quot;: [],
  &quot;interval_type&quot;: &quot;genes&quot;,
  &quot;default_level&quot;: 10,
  &quot;panel_name&quot;: &quot;Custom panel&quot;,
  &quot;case_display_name&quot;: &quot;internal_id&quot;,
  &quot;samples&quot;: [
    {
      &quot;name&quot;: &quot;string&quot;,
      &quot;coverage_file_path&quot;: &quot;string&quot;,
      &quot;case_name&quot;: &quot;string&quot;,
      &quot;analysis_date&quot;: &quot;2023-10-04T08:22:06.980106&quot;
    }
  ]
}
</code></pre>
<h3 id="mandatory-parameters">Mandatory parameters</h3>
<ul>
<li><strong>build</strong> <em>either "GRCh37" or "GRCh38"</em></li>
<li><strong>ensembl_gene_ids OR hgnc_gene_ids OR hgnc_gene_symbols</strong> <em>either a list of strings (for Ensembl IDs or HGNC symbols or a list of integers for HGNC IDs</em></li>
<li><strong>interval_type</strong> <em>either "genes", "transcripts" or "exons", depending on the analysis type (WGS, WES, WTS)</em></li>
<li><strong>default_level</strong> <em>default coverage threshold level to take into account for displaying the number of fully covered or partially covered intervals</em></li>
<li><strong>samples</strong> <em>mandatory parameters are <strong>name</strong> and either <strong>coverage_file_path</strong> or <strong>case_name</strong></em></li>
</ul>
<h3 id="optional-parameters">Optional parameters</h3>
<ul>
<li><strong>completeness_thresholds</strong> <em>the list of different coverage thresholds that will be used to calculate gene coverage completeness. If not provided, the default threshold values will be: 10, 15, 20, 50, 100</em></li>
<li><strong>panel_name</strong> <em>Add a name only if the provided genes are part of a gene panel</em></li>
<li><strong>case_display_name</strong> <em>provide this string parameter if the case belong to the same case / group</em></li>
<li><strong>samples analysis_date</strong> <em>if not provided with be set to current date</em></li>
</ul>
<h1 id="creating-a-genes-coverage-overview-report-to-show-incompletely-covered-genomic-intervals">Creating a genes coverage overview report to show incompletely covered genomic intervals</h1>
<p>This type of report contains stats from incompletely covered genomic intervals at different coverage thresholds and is basically the same as the genes overview report provided by <a href="https://github.com/Clinical-Genomics/chanjo-report/blob/9ca7a8ad279ab862c148ce4dfe2b2e675fdc4be4/chanjo_report/server/blueprints/report/views.py#L46">chanjo-report</a>.</p>
<p>A demo genes overview report based on genes transcripts from the demo PanelApp genes is available at the demo overview endpoint: http://0.0.0.0:8000/overview/demo (requires all genes and gene transcripts in build GRCh37 loaded into the database).</p>
<p><img width="1014" alt="image" src="https://github.com/Clinical-Genomics/chanjo2/assets/28093618/70bfdfb4-2345-47dd-b6e4-f6ea49d43cbc"></p>
<p>To create a custom genes coverage overview report, send a <strong>POST request to the <code>/overview</code> endpoint</strong> containing the same request data described above for the <code>/report</code> endpoint.</p>
<h1 id="creating-coverage-overview-over-mane-transcripts-for-a-list-of-genes">Creating coverage overview over MANE transcripts for a list of genes</h1>
<p><img width="1278" alt="image" src="https://github.com/user-attachments/assets/44eaa5c0-3777-42d0-8713-c33a2c5108cc"></p>
<p>This report contains statistics over all MANE Select and Mane Plus Clinical transcripts for a list of genes provided by the user. </p>
<p>The <strong><code>/mane_overview</code> endpoint</strong> accepts POST request with the same data described for the 2 reports above.</p>
<p>Note that MANE overview reports are available <ins>only for analyses run with genome build GRCh38</ins>.</p>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../sample-coverage/" class="btn btn-neutral float-left" title="Coverage and coverage completeness queries"><span class="icon icon-circle-arrow-left"></span> Previous</a>
        <a href="../authorised_requests/" class="btn btn-neutral float-right" title="Authorised requests">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
        <span>
          <a href="https://github.com/Clinical-Genomics/chanjo2/" class="fa fa-github" style="color: #fcfcfc"> GitHub</a>
        </span>
    
    
      <span><a href="../sample-coverage/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="../authorised_requests/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script src="../../js/jquery-3.6.0.min.js"></script>
    <script>var base_url = "../..";</script>
    <script src="../../js/theme_extra.js"></script>
    <script src="../../js/theme.js"></script>
      <script src="../../search/main.js"></script>
    <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script>

</body>
</html>
